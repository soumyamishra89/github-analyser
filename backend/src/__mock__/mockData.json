{
    "githubInfo": {
        "owner": "facebook",
        "name": "react",
        "requesterId": "2346",
        "url": "https://github.com/facebook/react",
        "commits": 12633,
        "openPullRequests": 264
    }
}