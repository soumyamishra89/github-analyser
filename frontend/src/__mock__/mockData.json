{
    "githubInfo": [{
        "owner": "facebook",
        "name": "react",
        "requesterId": "2346",
        "url": "https://github.com/facebook/react",
        "commits": 12633,
        "openPullRequests": 264
    },
    {
        "owner": "expressjs",
        "name": "express",
        "requesterId": "2134",
        "url": "https://github.com/expressjs/express",
        "commits": 5558,
        "openPullRequests": 62
    }]
}