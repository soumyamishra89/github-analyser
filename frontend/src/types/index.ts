export interface AppState {
    githubRepos: GithubRepo[],
    loggedStatus: LoggedStatus,
    isGithubReposLoading: boolean
}

export interface GithubRepo {
    owner: string,
    name: string,
    url: string,
    commits: number,
    openPullRequests: number,
    readme: string,
    requesterId: string, // id of the user who requested to fetch this data
    requestedOn: Date
}

export enum LoggedStatus {
    LOGGED_IN = 'LOGGEDIN',
    LOGGED_OUT = 'LOGGEDOUT'
}

export interface ReduxAction {
    type: ReduxActionType,
    data: any
}

export enum ReduxActionType {
    GITHUB_REPO = 'GITHUB_REPO',
    LOGGED_STATUS = 'LOGGED_STATUS',
    LOADING_GITHUB_REPO = 'LOADING_GITHUB_REPO'
}
