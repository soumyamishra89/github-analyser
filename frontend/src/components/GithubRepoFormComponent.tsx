import React from 'react';
import api from '../services/remote/api';
import { Form, Button, Col, Row } from 'react-bootstrap';
import { connect } from 'react-redux';
import { mapGithubDispatchToProps } from '../redux/dispatchMapping';
import { AppState } from '../types';

interface Props {
    loadGithubRepos: () => void,
    isGithubReposLoading: boolean;
}

interface State {
    owner: string,
    reponame: string,
    validated?: boolean,
    loading?: boolean
} 

export class GithubRepoFormComponent extends React.PureComponent<Props, State> {

    state: State = {
        owner: '',
        reponame: '',
        loading: false
    };

    componentDidUpdate(prevProps: Props) {
        if (prevProps.isGithubReposLoading === true && this.props.isGithubReposLoading === false) {
            this.setState({loading: false})
        }
    }

    postGithubInfo = async (event: any) => {
        event.preventDefault();
        const form = event.currentTarget;
        if (form.checkValidity() === false) {
            event.stopPropagation();
        } else {
            const { owner, reponame } = this.state;
            if (owner && reponame) {
                this.setState({loading: true});
                const resp = await api.postGithubInfo(owner, reponame);
                if (resp.status === 202) {
                    this.props.loadGithubRepos();
                } else {
                    this.setState({loading: false});
                    alert('Invalid github repo');
                }
            }
        }

        this.setState({validated: true});
    }

    onOwnerChange = (event: any) => {
        event.preventDefault();
        const value = event.target.value;
        this.setState({owner: value, validated: false});
    }

    onRepoChange = (event: any) => {
        event.preventDefault();
        const value = event.target.value;
        this.setState({reponame: value, validated: false});
    }

    render() {     
        
        return (
            <Form noValidate validated={this.state.validated} onSubmit={this.postGithubInfo}>
                <Form.Group>
                    <Form.Row>
                        <Col >
                            <Form.Control placeholder="Owner" required onChange={this.onOwnerChange} value={this.state.owner}/>
                        </Col>
                        <Col>
                            <Form.Control placeholder="Repo name" required onChange={this.onRepoChange} value={this.state.reponame}/>
                        </Col>
                        <Col>
                        <Button variant={this.state.loading ? "secondary" : "primary"} type="submit" className="position-relative d-flex justify-content-center align-items-center" disabled={this.state.loading}>
                            {this.state.loading && <div className="position-absolute button-loader"></div>}
                            Submit
                        </Button>
                        </Col>
                    </Form.Row>

                </Form.Group>
                <div className="d-flex justify-content-center">
                
                </div>
            </Form>
        )
    }
}

const mapStateToProps = (state: AppState) => ({
    isGithubReposLoading: state.isGithubReposLoading
});

export default connect(mapStateToProps, mapGithubDispatchToProps)(GithubRepoFormComponent);