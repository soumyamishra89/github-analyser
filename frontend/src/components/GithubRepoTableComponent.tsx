import React from 'react';
import { GithubRepo, AppState } from '../types';
import { Table, Modal } from 'react-bootstrap';
import { connect } from 'react-redux';
import Markdown from 'react-markdown';

interface Props {
    githubRepos: GithubRepo[]
}

interface State {
    readmeToShow?: string
}

class GithubRepoTableComponent extends React.PureComponent<Props, State> {

    state: State = {};

    showReadme = (evt: any, readme?: string) => {
        evt.preventDefault();
        if (!readme) {
            alert('No readme to show');
        } else {
            this.setState({readmeToShow: atob(readme)});
        }
    }

    hideReadme = () => {
        this.setState({readmeToShow: undefined});
    }

    render() {
        const { githubRepos } = this.props;
        if (githubRepos.length === 0) {
        return (<div className="d-flex justify-content-center align-item-center">
                Add a repository from github to show analysis
            </div>);
        }
        
        return (
            <>
                <table className="ml-2 mr-2 table table-striped table-bordered table-hover table-layout" >
                    <thead>
                        <tr>
                            <th>Owner</th>
                            <th>Repo name</th>
                            <th>Url</th>
                            <th>Commits</th>
                            <th>Open pull requests</th>
                            <th>Readme</th>
                        </tr>
                    </thead>
                    <tbody>
                        {githubRepos.map((gr, index) => (<tr key={index}>
                            <td>{gr.owner}</td>
                            <td>{gr.name}</td>
                            <td><a href={gr.url}>{gr.url}</a></td>
                            <td>{gr.commits}</td>
                            <td>{gr.openPullRequests}</td>
                            <td><button className="link-button" onClick={(evt: any) => this.showReadme(evt, gr.readme)}>Show Readme</button></td>
                        </tr>))}                    
                    </tbody>
                </table>
                {this.state.readmeToShow && <ReadmeModal onHide={this.hideReadme} readme={this.state.readmeToShow}/>}
            </>
        )
    }
}

interface ReadmeProps {
    onHide: () => void,
    readme: string
}

function ReadmeModal(props: ReadmeProps) {
    return (
        <Modal
            show={true}
            onHide={props.onHide}
            size="lg"
            aria-labelledby="contained-modal-title-vcenter"
            centered
        >
            <div className="pl-2 pr-2">
                <Markdown source={props.readme} />
            </div>
        </Modal>
    )
}

const mapStateToProps = (state: AppState) => ({
    githubRepos: state.githubRepos
});

export default connect(mapStateToProps)(GithubRepoTableComponent);