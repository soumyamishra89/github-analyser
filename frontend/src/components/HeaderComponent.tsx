import React from 'react';
import { Navbar, Nav } from 'react-bootstrap';
import { ThunkDispatch } from 'redux-thunk';
import { AppState } from '../types';
import PackageJson from '../../package.json';
import api from '../services/remote/api';
import { connect } from 'react-redux';
import GithubRepoFormComponent from './GithubRepoFormComponent';

interface Props {
    pingLoginStatus: () => void
}
class HeaderComponent extends React.PureComponent<Props, any> {

    componentDidMount() {
       this.props.pingLoginStatus();
    }

    render() {     
        
        return(
            <Navbar fixed="top" bg="dark" variant="dark" expand="lg">                   
                <Navbar.Brand>
                    Github Analyser <sup>{PackageJson.version}</sup>
                </Navbar.Brand>
                <Navbar.Toggle />
                <Navbar.Collapse>
                    <Nav className="mr-auto" />   
                    <GithubRepoFormComponent />
                </Navbar.Collapse>
            </Navbar>
        );
    }
}

const mapDispatchToProps = (dispatch: ThunkDispatch<AppState, any, any>) => ({
    pingLoginStatus: () => dispatch(api.pingLoginStatus())
})

export default connect(null, mapDispatchToProps)(HeaderComponent);