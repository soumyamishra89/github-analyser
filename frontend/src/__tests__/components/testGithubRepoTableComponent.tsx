import React from 'react';
import { ShallowWrapper, shallow } from 'enzyme';
import {GithubRepoTableComponent} from '../../components/GithubRepoTableComponent';
import mockData from '../../__mock__/mockData.json';
import { GithubRepo } from '../../types';

const createProps = () => ({
    githubRepos: (mockData.githubInfo as GithubRepo[])
})

describe("Test GithubRepoTableComponent rendering", () => {
    let props;
    let wrapper: ShallowWrapper;
    beforeEach(() => {
        props = createProps();
        
        wrapper = shallow(<GithubRepoTableComponent {...props}/>);
    });

    it("Snapshot test with data", () => {
        expect(wrapper).toMatchSnapshot("With data");
    });

    it("Snapshot test without data", () => {
        wrapper.setProps({githubRepos: []});
        expect(wrapper).toMatchSnapshot("Without data");
    });
})