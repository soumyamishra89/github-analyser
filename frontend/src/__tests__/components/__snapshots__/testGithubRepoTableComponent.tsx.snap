// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test GithubRepoTableComponent rendering Snapshot test with data: With data 1`] = `
<Fragment>
  <table
    className="ml-2 mr-2 table table-striped table-bordered table-hover table-layout"
  >
    <thead>
      <tr>
        <th>
          Owner
        </th>
        <th>
          Repo name
        </th>
        <th>
          Url
        </th>
        <th>
          Commits
        </th>
        <th>
          Open pull requests
        </th>
        <th>
          Readme
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="0"
      >
        <td>
          facebook
        </td>
        <td>
          react
        </td>
        <td>
          <a
            href="https://github.com/facebook/react"
          >
            https://github.com/facebook/react
          </a>
        </td>
        <td>
          12633
        </td>
        <td>
          264
        </td>
        <td>
          <button
            className="link-button"
            onClick={[Function]}
          >
            Show Readme
          </button>
        </td>
      </tr>
      <tr
        key="1"
      >
        <td>
          expressjs
        </td>
        <td>
          express
        </td>
        <td>
          <a
            href="https://github.com/expressjs/express"
          >
            https://github.com/expressjs/express
          </a>
        </td>
        <td>
          5558
        </td>
        <td>
          62
        </td>
        <td>
          <button
            className="link-button"
            onClick={[Function]}
          >
            Show Readme
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</Fragment>
`;

exports[`Test GithubRepoTableComponent rendering Snapshot test without data: Without data 1`] = `
<div
  className="d-flex justify-content-center align-item-center"
>
  Add a repository from github to show analysis
</div>
`;
